### PR TITLE
Issue 2 -  General Details Screen

### LINKED ISSUE #3: General Details Screen
MicrosoftTeams-image (44)

### PR DESCRIPTION
**Pull request by PreetiRajguru**
- Implemented Stepper 
- Created General Details Component
- Added Dynamic Data to the Component (Data, Links)
- Added Form Validations (1.Validation on Empty Input, 2.Validation on Next Button)
- Css Enhancements (Conditional Rendering)
- Lifted General Details Component State up to another Component to handle Validations

### CODE REVIEW

File: connectors/src/Components/Configure.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+const steps = [
+    { label: 'General Details', component: GeneralDetails },
+    { label: 'Connect Account', component: ConnectAccount },
+    { label: 'Sync & Mappings', component: SyncMappings },
+    { label: 'Filters', component: ConnectAccount },
+    { label: 'Test & enable Connection', component: GeneralDetails },
+];
+
+function DoneIcon() {
+    return "✓";
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js]: Can this be moved to Assets? if this is a icon/img?

File: connectors/src/Components/Configure.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import React, { useState } from 'react';
+import GeneralDetails from './GeneralDetails';
+import ConnectAccount from './ConnectAccount';
+import SyncMappings from './SyncMappings';
+import { Link } from 'react-router-dom';
+
+const steps = [
+    { label: 'General Details', component: GeneralDetails },
+    { label: 'Connect Account', component: ConnectAccount },
+    { label: 'Sync & Mappings', component: SyncMappings },
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js]: Incorrect component rendering. GenralDetails & ConnectAccount component are rendered twice, I understand that this is just for testing purpose but it should removed while creating the PR

File: connectors/src/Components/Configure.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                <div className="stepper__step__content">
+                    <div className="stepper-step-index step-circle stepper-index"><StepContent done={done} index={key} /></div>
+                    <div className='stepper-labels'>{label}{icon}</div>
+                </div>
+            </div>
+        </li>
+    )
+}
+
+function Stepper({ current, steps }) {
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js:45]: Not sure what is the standard way of declaring the function name but either of `Stepper` or 'renderStep' is incorrect

File: connectors/src/Components/Configure.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                generalDetails={generalDetails}
+                setGeneralDetails={setGeneralDetails}
+                validationErrors={validationErrors}
+                setValidationErrors={setValidationErrors}
+            />
+
+            <div>
+                <hr className='bottom-hr'></hr>
+                <Link to="/" className="btn btn-outline-secondary me-2 cancel-button">Cancel </Link>
+                <button type="button" className={saveAndFinishClass}>Save & Finish</button>
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js]: I believe this could have been a link

File: connectors/src/Components/GeneralDetails.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import React from 'react';
+
+const GeneralDetails = ({ generalDetails, setGeneralDetails, validationErrors, setValidationErrors }) => {
+
+    const GeneralDetailsData = [{
+        desc: "Microsoft Dynamics is a line of enterprise Resource Planning and customer relationship management software applications. Microsoft markets Dynamics applications through a network of reselling partners who provide specialized services. ",
+        link1: '#',
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/GeneralDetails.js]: What is link1, link2?

File: connectors/src/Components/GeneralDetails.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+            {GeneralDetailsData.map((item, index) => {
+                return (
+                    <div key={index}>
+
+                        <span>
+                            {item.desc}
+                        </span>
+
+                        <br /><br />
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/GeneralDetails.js]: We should avoid <br>

File: connectors/src/Components/Sidebar.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+        {/* Sidebar content */}
+        <div className='sidebar-new'>
+          <i class="bi bi-gear-fill sidebar-icon settings-headers"></i>
+          <span className='settings-headers settings-header'>SETTINGS</span>
+        </div>
+
+        <hr className='hr-tag'></hr>
+
+        <ul className='sidebar-list'>
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Sidebar.js:57]: These menu's could have been rendered dynamically. Imagine if we have to reduce these menu's

**Review by RavirajSuryavanshi — CHANGES_REQUESTED** [review]

File: connectors/src/Components/Configure.js (hunk 2 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js]: Used GeneralDetails Component for rendering . Removed the use of ConnectAccount and SyncMappings.

File: connectors/src/Components/Configure.js (hunk 3 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js:45]: All these are Functional Components used for implementing Stepper so they follow Pascal Case Name Convention.
Changed 'renderStep' to 'RenderStep'.

File: connectors/src/Components/Configure.js (hunk 1 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js]: Added Bootstrap Icon

File: connectors/src/Components/Configure.js (hunk 4 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js]: Added <Link>

File: connectors/src/Components/GeneralDetails.js (hunk 5 above)
Comment (PreetiRajguru) [reply connectors/src/Components/GeneralDetails.js]: Added Meaningful Names to the Links

File: connectors/src/Components/GeneralDetails.js (hunk 6 above)
Comment (PreetiRajguru) [reply connectors/src/Components/GeneralDetails.js]: Removed Usage of Br Tags

File: connectors/src/Components/Sidebar.js (hunk 7 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Sidebar.js:57]: Sidebar component made Dynamic

File: connectors/src/Components/Configure.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    const handlePrevious = () => {
+        setCurrent(Math.max(current - 1, 0));
+    }
+
+    const isStep1 = current === 0;
+    const saveAndFinishClass = isStep1 ? 'btn btn-link save-and-finish-button' : 'btn btn-link save-and-finish-button2';
+
+    return (
+        <>
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js:75]: What is the use of <> ?? Something new with react syntax??

File: connectors/src/Components/Configure.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+        setCurrent(Math.min(current + 1, steps.length));
+    }
+
+    const handlePrevious = () => {
+        setCurrent(Math.max(current - 1, 0));
+    }
+
+    const isStep1 = current === 0;
+    const saveAndFinishClass = isStep1 ? 'btn btn-link save-and-finish-button' : 'btn btn-link save-and-finish-button2';
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js]: Please find a better way to apply conditional styling

File: connectors/src/Components/Configure.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+function StepperState() {
+    const [current, setCurrent] = useState(0);
+    const [generalDetails, setGeneralDetails] = useState({});
+    const [validationErrors, setValidationErrors] = useState({});
+    const CurrentStepComponent = steps[current].component;
+
+    const handleNext = () => {
+
+        if (!generalDetails.name) {
+            setValidationErrors({ generalDetails: "Name is required." });
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Configure.js]: Error messages should not be hardcoded

File: connectors/src/Components/Navbar.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                        <button type="button" class="btn btn-outline-warning" id="ro-button">
+                            RO
+                            <span id="orange-dot"></span>
+                        </button>
+                        <span id="vertical-line-buttons">|</span>
+
+                        <button type="button" class="btn btn-outline-primary" id="call-button">
+                            <i class="bi bi-telephone-plus-fill"></i>
+                        </button>
+
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Navbar.js:34]: Too many line spaces. Please take of this we should not be discussing on this multiple times.

File: connectors/src/Components/Navbar.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+                                    type="button"
+                                    className="btn btn-secondary dropdown-toggle"
+                                    data-toggle="dropdown"
+                                    aria-haspopup="true"
+                                    aria-expanded="false"
+                                >
+                                    <i class="bi bi-gear-fill search-color"></i>
+                                    <i class="bi bi-caret-down-fill caret-down caret-down"></i>
+                                </button>
+
```
Comment (RavirajSuryavanshi) [inline connectors/src/Components/Navbar.js:50]: Too many line spaces. Please take of this we should not be discussing on this multiple times.

File: connectors/src/style.css [hunk 13]
```diff
@@ -403,4 +409,227 @@
...
+.step-circle {
+    margin-left: -114px
+}
+
+.input-icon {
+    position: relative;
+}
+
+.input-icon-icon {
+    /* position: absolute; */
```
Comment (RavirajSuryavanshi) [inline connectors/src/style.css]: Remove commented code. Please take of this we should not be discussing on this multiple times.

File: connectors/src/style.css [hunk 14]
```diff
@@ -403,4 +409,227 @@
...
+
+.cancel-button {
+    margin-left: 25px;
+    margin-top: 60px
+}
+
+.save-and-finish-button {
+    margin-top: 60px;
+    margin-left: 77rem;
+    text-decoration: none !important
```
Comment (RavirajSuryavanshi) [inline connectors/src/style.css]: Important Why??

**Review by RavirajSuryavanshi — CHANGES_REQUESTED** [review]

File: connectors/src/Components/Configure.js (hunk 8 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js:75]: - <> are called "Fragments." 
- Used to group multiple child elements
- Can be replaced with extra  <div> division tags.

File: connectors/src/Components/Navbar.js (hunk 11 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Navbar.js:34]: Removed Extra Spaces.

File: connectors/src/Components/Navbar.js (hunk 12 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Navbar.js:50]: Removed Extra Spaces.

File: connectors/src/style.css (hunk 13 above)
Comment (PreetiRajguru) [reply connectors/src/style.css]: Removed all Commented Code

File: connectors/src/Components/Configure.js (hunk 10 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js]: Added ConstantMessages.js file.

File: connectors/src/Components/Configure.js (hunk 9 above)
Comment (PreetiRajguru) [reply connectors/src/Components/Configure.js]: Added Inline Conditional Rendering of  CSS and removed the usage of an extra variable for the same.

File: connectors/src/style.css (hunk 14 above)
Comment (PreetiRajguru) [reply connectors/src/style.css]: Removed Usage of !important

**Comment by RavirajSuryavanshi** [comment]
Merging PR as all comments have been resolved.

**Pull request merged**